File: ta_lib/strategies/regime_filter/src/rsi.rs
use base::{Regime, OHLCVSeries};
use core::Series;
use shared::{rsi_indicator, RSIType};

const RSI_NEUTRALITY: f32 = 50.0;

pub struct RSIFilter {
    rsi_type: RSIType,
    period: usize,
    threshold: f32,
}

impl RSIFilter {
    pub fn new(rsi_type: RSIType, period: f32, threshold: f32) -> Self {
        Self {
            rsi_type,
            period: period as usize,
            threshold,
        }
    }
}

impl Regime for RSIFilter {
    fn lookback(&self) -> usize {
        self.period
    }

    fn apply(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let rsi = rsi_indicator(&self.rsi_type, data, self.period);

        (
            rsi.sgt(RSI_NEUTRALITY + self.threshold),
            rsi.slt(RSI_NEUTRALITY - self.threshold),
        )
    }
}

File: strategy/generator/filter/rsi.py
from dataclasses import dataclass

from core.models.parameter import CategoricalParameter, Parameter, RandomParameter
from core.models.rsi import RSIType

from .base import Filter, FilterType


@dataclass(frozen=True)
class RsiFilter(Filter):
    type: FilterType = FilterType.Rsi
    rsi_type: Parameter = CategoricalParameter(RSIType)
    period: Parameter = RandomParameter(30.0, 50.0, 1.0)
    threshold: Parameter = RandomParameter(2.0, 5.0, 1.0)
